### ISSUE TITLE
Sporadic errors "LoadError: no such file to load -- sass" with JRuby and Sprockets

### ISSUE THREAD

**Issue by chadlwilson** [ntkme/sass-embedded-host-ruby#64]
I'm not really sure what is going on here, but would appreciate any tips - please go gentle, I'm not a Rails/Sprockets expert 😅. Apologies if this should be reported at the other project; I'm not really sure.

When working in `development` mode with hot-reloading and the autoloader, every now and then (but reasonably frequently) I get the below error. I can't work out a pattern as to what is happening. When this happens the browser will block. If I kill the request and try again it will usually work OK but then occur again later.

I have also had it happen when pre-compiling assets too, so perhaps it is to do with some kind of race condition or threading type of issue with JRuby or sprockets? If you suspect I need to upgrade to Sprockets 4, that's fine. Haven't noticed a similar issue earlier when using ruby-sass.

```
warning: thread "Ruby-0-Thread-4163: /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/bundler/gems/sass-embedded-host-ruby-0d103fd1936a/lib/sass/embedded/dispatcher.rb:18" terminated with exception (report_on_exception is true):
LoadError: no such file to load -- sass
                require at org/jruby/RubyKernel.java:1017
                require at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/gems/zeitwerk-2.6.0/lib/zeitwerk/kernel.rb:35
                 <main> at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/gems/sprockets-3.7.2/lib/sprockets/autoload/sass.rb:1
                require at org/jruby/RubyKernel.java:1017
                require at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/gems/zeitwerk-2.6.0/lib/zeitwerk/kernel.rb:35
              asset_url at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/gems/sprockets-3.7.2/lib/sprockets/sass_processor.rb:137
               font_url at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/gems/sprockets-3.7.2/lib/sprockets/sass_processor.rb:209
                  setup at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/gems/sassc-embedded-1.54.0/lib/sassc/embedded.rb:151
          function_call at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/bundler/gems/sass-embedded-host-ruby-0d103fd1936a/lib/sass/embedded/host/function_registry.rb:50
  function_call_request at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/bundler/gems/sass-embedded-host-ruby-0d103fd1936a/lib/sass/embedded/host.rb:119
            public_send at org/jruby/RubyKernel.java:2003
        receive_message at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/bundler/gems/sass-embedded-host-ruby-0d103fd1936a/lib/sass/embedded/dispatcher.rb:80
```

**Environment**:
JRuby 9.3.7.0 (Ruby 2.6 compatible)
Rails 6.1.6.1 in Zeitwork mode
Sprockets 3.7.2 (haven't yet moved to 4.1, was earlier blocked by another dependency)

**Comment by ntkme** [comment]
Sounds like a duplicate of https://github.com/ntkme/sassc-embedded-shim-ruby/issues/7

**Comment by ntkme** [comment]
Assuming you’re trying to use the shim for `sassc`, you will have to make sure all Sass related configs are referring `sassc` rather than `sass` or you may see this error.

**Comment by ntkme** [comment]
Just checked, you do need Sprockets 4.x for sassc support.

**Comment by chadlwilson** [comment]
Hmm, that does seem similar - however doesn't seem to explain why it's mostly working just fine, but occasionally fails? We're not overriding any defaults for the `css_compressor` at all right now.

I did have `sassc` working earlier with Sprockets 3.7.x as well before switching to sass-embedded due to issues with Windows.

If you think there's something in Sprockets 4 likely required for it to be better behaved here, I suppose I'll have to try and upgrade that too. 😬

**Comment by ntkme** [comment]
You can see SassC was not supported in `3.x` by comparing `3.x` branch and `main`: https://github.com/rails/sprockets/compare/3.x...main#diff-7f52931473a0b96f437132b1a55147137bba863f9f65e5824e19ab71bd0a3028R11

When you say mostly working, it is probably due to the lazy loading (a.k.a. `autoload`) of libraries only when needed.

**Comment by ntkme** [comment]
Here is my educated guess: you have `sassc` installed, but it was never used by `sprockets 3.x`. You probably have `sass` somehow available in Ruby's `$LOAD_PATH` like the reporter of the other issue, so it happens to work sometimes. E.g. `rake` and `bundle exec rake` have different `$LOAD_PATH`.

I don't know. All I can tell is that it is not supposed to work at all `sprockets 3.x`.

**Comment by chadlwilson** [comment]
Ok, thanks a lot. Pretty sure I don't have `sass` available anywhere, but regardless will try again with Sprockets 4. 🙏

**Comment by chadlwilson** [comment]
For what it's worth, I'm still able to get this on Sprockets 4, but ran out of time to figure out what is going on. Somehow the `sass_processor` is being used rather than sassc_processor` at times.

```
warning: thread "Ruby-0-Thread-1672: /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/bundler/gems/sass-embedded-host-ruby-0d103fd1936a/lib/sass/embedded/dispatcher.rb:18" terminated with exception (report_on_exception is true):
LoadError: no such file to load -- sass
                require at org/jruby/RubyKernel.java:1017
                require at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/gems/zeitwerk-2.6.0/lib/zeitwerk/kernel.rb:35
                 <main> at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/gems/sprockets-4.1.1/lib/sprockets/autoload/sass.rb:2
                require at org/jruby/RubyKernel.java:1017
                require at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/gems/zeitwerk-2.6.0/lib/zeitwerk/kernel.rb:35
              asset_url at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/gems/sprockets-4.1.1/lib/sprockets/sass_processor.rb:158
              image_url at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/gems/sprockets-4.1.1/lib/sprockets/sass_processor.rb:176
                  setup at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/gems/sassc-embedded-1.54.0/lib/sassc/embedded.rb:151
          function_call at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/bundler/gems/sass-embedded-host-ruby-0d103fd1936a/lib/sass/embedded/host/function_registry.rb:50
  function_call_request at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/bundler/gems/sass-embedded-host-ruby-0d103fd1936a/lib/sass/embedded/host.rb:119
            public_send at org/jruby/RubyKernel.java:2003
        receive_message at /Users/chad/Projects/community/gocd/gocd/server/src/main/webapp/WEB-INF/rails/gems/jruby/2.6.0/bundler/gems/sass-embedded-host-ruby-0d103fd1936a/lib/sass/embedded/dispatcher.rb:80
```

**Comment by ntkme** [comment]
I think the issue has to do with when this line is executed: https://github.com/sass/sassc-rails/blob/8d0462d54b5b5dd84cb1df31823c3afaebb64534/lib/sassc/rails/functions.rb#L13

**Comment by ntkme** [comment]
https://github.com/sass/sassc-rails/issues/114 seems to be related here.

**Comment by ntkme** [comment]
@chadlwilson This is probably it: https://github.com/sass/sassc-rails/issues/114#issuecomment-441185698
